hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -940,18 +932,61 @@ private boolean shouldMergeTo(Address thisAddress, Address targetAddress) {
     }
 
     void reset() {
-        clusterServiceLock.lock();
-        try {
-            joinInProgress = false;
-            joiningMembers.clear();
-            timeToStartJoin = Clock.currentTimeMillis() + waitMillisBeforeJoin;
-            firstJoinRequest = 0;
-        } finally {
-            clusterServiceLock.unlock();
+        if (syncJoinStrategy == null) {
+            if (cancelMigrationTimeout()) {
+                node.getPartitionService().resumeMigration();
+            }
+            mastershipClaimInProgress = false;
+        } else {
+            clusterServiceLock.lock();
+            try {
+                joinInProgress = false;
+                syncJoinStrategy.reset();
+            } finally {
+                clusterServiceLock.unlock();
+            }
         }
     }
 
+    /**
+     * assumes clusterServiceLock is locked
+     *
+     * @return true if timeout needed to be canceled
+     */
+    private boolean cancelMigrationTimeout() {
+        boolean timedOut = migrationDelayActive.getAndSet(false);
+        if (timedOut) {
+            minDelayFuture.cancel(false);
+            maxDelayFuture.cancel(false);
+            // only for posterity's sake
+            minDelayFuture = null;
+            maxDelayFuture = null;
+        }
+        return timedOut;
+    }
+
+    /**
+     * assumes clusterServiceLock is locked
+     */
+    private void scheduleMigrationDelay() {
+        boolean subsequentJoinAttempt = migrationDelayActive.getAndSet(true);
+        if (subsequentJoinAttempt) {
+            assert minDelayFuture.cancel(false) : "Something went wrong canceling min delay future";
+        }
+        minDelayFuture = nodeEngine.getExecutionService().schedule(this::reset,
+                waitMillisBeforeJoin, TimeUnit.MILLISECONDS);
+        if (!subsequentJoinAttempt) {
+            // pause migrations until no more members are trying to join in the same period
+            node.getPartitionService().pauseMigration();
```

Comment:
Since the call to this method and to `reset` are concurrent, the migration might be resumed as soon as it's paused. I also am not sure of the effects of scheduling the call to reset. It may happen that by the time the reset method is called, mastership claim is in progress, or some join is in progress. The reset method will then reset the state of this node. Is that correct?

In another words, the reset method was called in very specific times - when this node is joining another cluster, when it's merging from a split-brain, when it claimed mastership or when it was shutting down. Now, it's scheduled and I believe this opens up to some race conditions.

Reply:
all these methods are called as a result of AtomicBoolean.getAndSet() so it's only getting called when the state changes from false to true or vice-versa, which eliminates the race confitions

Reply:
Also the mastership claim process in this case works exactly the same w/ concurrent calls to reset as in prior version. The only thing hanged w/mastership process is that instead of using the same variable (`joinInProgress` / non-empty map), the more correct way is used now by introducing a second variable